Modix.Bot/Behaviors/PromotionPollBehavior/PromotionDialogueBehavior.cs

```diff
@@ -0,0 +1,161 @@
+        private readonly DateTime _utcNow = DateTime.UtcNow;
+
+        private static readonly Emoji _approveEmoji = new Emoji("👍");
+
+        private static readonly Emoji _disproveEmoji = new Emoji("👎");
+
+        private IDesignatedChannelService _designatedChannelService { get; }
+
+        private IDiscordSocketClient _discordSocketClient { get; }
+
+        private IMemoryCache _memoryCache { get; }
+
+        private IPromotionsService _promotionsService { get; }
+
+        private IImageService _imageService { get; }
+
+        public PromotionDialogueBehavior(IDesignatedChannelService designatedChannelService,
+            IDiscordSocketClient discordSocketClient,
+            IMemoryCache memoryCache,
+            IPromotionsService promotionsService,
+            IImageService imageService)
+
+        {
+            _designatedChannelService = designatedChannelService;
+            _discordSocketClient = discordSocketClient;
+            _memoryCache = memoryCache;
+            _promotionsService = promotionsService;
+            _imageService = imageService;
+        }
+
+        public async Task HandleNotificationAsync(PromotionActionCreatedNotification notification,
+            CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken.IsCancellationRequested)
+                return;
+
+            switch (notification.Data.Type)
+            {
+                case PromotionActionType.CampaignCreated:
+                    await CreatePromoPoll(notification.Data);
+                    break;
+                case PromotionActionType.CampaignClosed:
+                    await DeletePromoPoll(notification.Data);
+                    break;
+            }
+        }
+
+        public async Task HandleNotificationAsync(ReactionAddedNotification notification,
+            CancellationToken cancellationToken = default)
+        {
+            var reaction = notification.Reaction.Emote;
+            var user = (IGuildUser)notification.Reaction.User.Value;
+
+            if (cancellationToken.IsCancellationRequested
+                || notification.Reaction.User.Value.IsBot
+                || notification.Message.Value.Embeds.Count < 1)
+                return;
+
+            if (_memoryCache.TryGetValue(notification.Message.Id, out CachedPromoPoll poll)
+            && ValidateVoteReaction(reaction))
+            {
+                await notification.Message.Value.RemoveReactionAsync(notification.Reaction.Emote, notification.Reaction.User.Value);
+
+                var sentiment = reaction.Name == _approveEmoji.Name ? PromotionSentiment.Approve : PromotionSentiment.Oppose;
+
+                if (!await _promotionsService.VotePromoAsync(poll.Campaign.Id, user,  sentiment))
+                    return;
+
+                if (reaction.Name == _approveEmoji.Name)
+                    ++poll.Approve;
+                else if(reaction.Name == _disproveEmoji.Name)
+                    ++poll.Disprove;
```

Comment:
I would prefer a better way of doing this as this feels hacky

Reply:
Not sure what you mean. Seems fine, except I would definitely make `CachedPromoPoll` an immutable object, where you create a new one and update it into the cache. 

Alternatively, just re-query the DB each time the campaign is updated, so you can never end up with "wrong" data.

Alternatively alternatively, don't screw with the dialog at all here. Let the promotions system publish the proper notifications for when a comment is created or modified or whatever, and just handle those notifications.

Reply:
I like the notification way, adds another handler which is imo clearer

Reply:
:) Me too.